### ISSUE TITLE
Depend on ConstructionBase.jl to allow funcitonal updating of quantities etc.

### ISSUE THREAD

**Issue by rafaqz** [PainterQubits/Unitful.jl#277]
Most of the types in Unitful.jl can't be determined from their fields, so rebuilding them with
Setfield.jl or Flatten.jl doesn't work by default. But these tools can be very useful for automaticaly separating out unitful quantities from optimisers etc that need simple AbstractFloat.

Flatten.jl used to just do this internally, but now ConstructionBase provides `constructorof` method, it would be type pyracy to define it for `Quantity`.

Does anyone object to a dependency on [ConstructionBase.jl](https://github.com/JuliaObjects/ConstructionBase.jl)? Its tiny and has no deps.

**Comment by ajkeller34** [comment]
> types in Unitful.jl can't be determined from their fields, so rebuilding them with
Setfield.jl or Flatten.jl doesn't work by default.

Can you please elaborate? I'm not sure what problem you are trying to solve here and am unfamiliar with these packages.

> these tools can be very useful for automaticaly separating out unitful quantities from optimisers etc that need simple AbstractFloat.

I think you're getting at the fact that some packages are not written generically enough to use Unitful quantities, and I would be curious to see your approach to remedying that situation, since you seem to have some examples in mind.

> Flatten.jl used to just do this internally, but now ConstructionBase provides `constructorof` method, it would be type pyracy to define it for `Quantity`.

Please provide more context, I'm guessing there are many others besides myself who are a bit lost here.

> Does anyone object to a dependency on ConstructionBase.jl? Its tiny and has no deps.

For a long time I have been able to avoid adding dependencies to Unitful, and am unlikely to add a dependency now. I don't mean to be dogmatic and I could be convinced so long as there are very good reasons, but it seems at first glance like ConstructionBase is very new and a bit experimental.

**Comment by rafaqz** [comment]
Ok I'll try to describe this a little better. With my main use-case for Flatten.jl, I write packages that accept arbitrary models, that may or may not use units. But I dont know where the units are if they are used by users.

Using Flatten.jl any model structure (nested structs/tuples etc) can be reduced to a tuple/vector or `Real`, to be sent an optimiser without hardcoding the transformation. Which is a powerful generalisation. Models can also be updated to hold different types like dual numbers, or swap out float types etc.

But this breaks with units, because the constructor for Quantity has D and U parameters that can't be determined by the field type. In some cases the type of Quantity has to change ie with dual numbers or when conversion to array has promoted float types, so you can't always use the original constructor.

To get around this you have to define a constructor for Quantity, like:
```julia
Unitful.Quantity{<:Any, D, U}(val) where {D, U} = Quantity{typeof(val / unit(val)), D, U}(val)

constructorof(::Type{Unitful.Quantity{_, D, U}}) where {_, D, U} =
    Unitful.Quantity{T, D, U} where T
```
This allows you to swap out the val field of Quantity without knowing you are dealing with Quantity at all, its just a generic transformation of an object using its fields and `constructorof`.

ConstructionBase generalises this so that Setfield and other similar packages can all use this constructor.

So you can do:

```julia
julia> u = 1u"m"
1 m

julia> @set u.val = 2.0
2.0 m
```
But this is a trivial example - Setfield defines lenses that let you define many kinds of functional updates to immutable types that you can automate at larger scales.

ConstructorOf isn't really experimental, it barely does anything besides define the methods `constructorof` and `setproperties` (without the bang). But I can relate to resisting the dependency. My preference was to automate the generation of these constructors, but it turns out to be a difficult thing to do and very fragile given how complicated type definitions can get.

**Comment by rafaqz** [comment]
This is what flatten currently does, using Requires.jl to define a constructor for Unitful.jl `Quantity` when it is loaded:
https://github.com/rafaqz/Flatten.jl/blob/master/src/unitful.jl

But that can't be used for Setfield.jl or BangBang.jl, and you would need to define it again for those packages.

**Comment by cstjean** [comment]
I don't have anything meaningful to add, but FWIW I've had similar issues in the past, and I believe that the `ConstructionBase` developments are really cool for machine learning using complex models.

**Comment by rafaqz** [comment]
An option without any dependencies is to change the types so that all type parameters have matching fields in the same order. Then construction would work automatically. 

I'm not sure if that would break the API in any way.

**Comment by ajkeller34** [comment]
> An option without any dependencies is to change the types so that all type parameters have matching fields in the same order. Then construction would work automatically.
> 
> I'm not sure if that would break the API in any way.

It's a good idea, but the dimension type parameter of a `Quantity` is actually an object, not a type, so I don't think it would work. This was done a while ago so that when the type of a `Quantity` is printed, it isn't quite as long as it otherwise would be (which is saying something, because sometimes they are very long anyhow). 

I'm warming to the idea of depending on `ConstructionBase`. My main worry is some subtle issues creeping in owing to the use of generated functions, having been bitten by that in Unitful itself.

**Comment by rafaqz** [comment]
Ok that makes sense.

I would hope the `@generated` functions would only be run when someone uses another package that depends on ConstructionBase and calls `constructorof`, which would partition that risk off to people that are taking it anyway. But of course I don't really know how the compiler works lol.

**Comment by ajkeller34** [comment]
Closed by #280
